File: python/tickbridge.py
import asyncio
import os
import json
from tickengine_sdk import TickEngineClient
from agents import Mt5Agent, BinanceAgent, OkxAgent


def _order_type_raw(val) -> int:
    """Map stream ClientOrderType string to MQL5 int. 0=Market 1=Limit 2=Stop."""
    if val in (1, "Limit", "limit"): return 1
    if val in (2, "Stop",  "stop"):  return 2
    return 0


def _order_type_label(raw: int) -> str:
    return ["MARKET", "LIMIT", "STOP"][raw]


def _side(val) -> tuple[str, int]:
    is_buy = val in (0, "Buy", "buy")
    return ("BUY" if is_buy else "SELL", 0 if is_buy else 1)


def extract_order_details(event: dict) -> dict | None:
    etype = event.get("type")
    data  = event.get("data", {})

    if etype == "trade":
        # TradeExecutedEventDto (camelCase): type → "Market"/"Limit"/"Stop", tradeId, side, size, price
        ot = _order_type_raw(data.get("type"))
        sd, sd_raw = _side(data.get("side"))
        signal_id = data.get("entryId") if data.get("status") == "closed" and data.get("entryId") else data.get("tradeId")
        return {
            "symbol":         data.get("symbol"),
            "side":           sd,
            "order_type":     _order_type_label(ot),
            "quantity":       float(data.get("size", 0)),
            "price":          float(data.get("price", 0)),
            "signal_id":      signal_id,
            "timestamp":      int(data.get("timestamp", 0)),
            "event_type":     0,
            "order_type_raw": ot,
            "side_raw":       sd_raw,
        }

    if etype == "order":
        # OrderEventDto (camelCase): orderType, triggerPrice, orderId, side, size
        ot = _order_type_raw(data.get("orderType"))
        sd, sd_raw = _side(data.get("side"))
        return {
            "symbol":         data.get("symbol"),
            "side":           sd,
            "order_type":     _order_type_label(ot),
            "quantity":       float(data.get("size", 0)),
            "price":          float(data.get("triggerPrice") or 0),
            "signal_id":      data.get("orderId"),
            "timestamp":      int(data.get("timestamp", 0)),
            "event_type":     1,
            "order_type_raw": ot,
            "side_raw":       sd_raw,
        }

    if etype == "alert":
        # WatchAlert: no side/orderType by default — fields are optional user extensions
        ot = _order_type_raw(data.get("orderType") or data.get("order_type", "Market"))
        sd, sd_raw = _side(data.get("side", "buy"))
        return {
            "symbol":         data.get("symbol", "global"),
            "side":           sd,
            "order_type":     _order_type_label(ot),
            "quantity":       float(data.get("size", 0)),
            "price":          float(data.get("price", 0)),
            "signal_id":      data.get("id"),
            "timestamp":      int(data.get("timestamp", 0)),
            "event_type":     2,
            "order_type_raw": ot,
            "side_raw":       sd_raw,
        }

    return None


async def main():
    config_file = os.getenv("TICKENGINE_CONFIG_FILE", "config.json")
    print(f"Loading config from: {config_file}")
    with open(config_file, "r") as f:
        config = json.load(f)

    agents = []
    for agent_name, agent_cfg in config.get("agents", {}).items():
        atype      = agent_cfg.get("type")
        symbol_map = agent_cfg.get("symbol_map", {})

        if atype == "mt5_ea":
            try:
                agent = Mt5Agent(agent_name, symbol_map, agent_cfg.get("tcp_bind"))
                await agent.start()
                agents.append(agent)
            except Exception as e:
                print(f"Failed to initialize MT5 Agent '{agent_name}': {e}")
        elif atype == "binance_spot":
            agents.append(BinanceAgent(
                name=agent_name, symbol_map=symbol_map,
                api_key=agent_cfg.get("api_key"), api_secret=agent_cfg.get("api_secret"),
                api_url=agent_cfg.get("api_url", "https://api.binance.com"),
            ))
        elif atype == "okx_spot":
            agents.append(OkxAgent(
                name=agent_name, symbol_map=symbol_map,
                api_key=agent_cfg.get("api_key"), api_secret=agent_cfg.get("api_secret"),
                passphrase=agent_cfg.get("passphrase"),
                api_url=agent_cfg.get("api_url", "https://www.okx.com"),
            ))

    client = TickEngineClient(
        base_url=config["stream"]["url"],
        api_key=config["stream"]["api_key"],
        account_id=config["stream"]["account_id"],
    )

    print("Connecting to TickEngine stream...")
    async for event in client.stream_events():
        details = extract_order_details(event)
        if not details:
            continue
        for agent in agents:
            if agent.can_handle(details["symbol"]):
                try:
                    await agent.execute(event, details)
                except Exception as e:
                    print(f"Error executing event via Agent '{agent.name}': {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: nodejs/tickbridge.ts
import * as fs from "fs";
import { TickEngineClient } from "./index";
import { Mt5Agent, BinanceAgent, OkxAgent, OrderDetails, ExecutionAgent } from "./agents";

// ---------------------------------------------------------------------------
// Order type helpers
// ---------------------------------------------------------------------------

/** Maps stream order type value → MQL5 raw int. 0=Market, 1=Limit, 2=Stop */
function orderTypeRaw(val: any): number {
    if (val === 1 || val === "Limit" || val === "limit") return 1;
    if (val === 2 || val === "Stop"  || val === "stop")  return 2;
    return 0; // Market (default)
}

function orderTypeLabel(raw: number): string {
    if (raw === 1) return "LIMIT";
    if (raw === 2) return "STOP";
    return "MARKET";
}

// ---------------------------------------------------------------------------
// Event parsing  (stream serialises all DTOs with rename_all = "camelCase")
// ---------------------------------------------------------------------------

function extractOrderDetails(event: any): OrderDetails | null {
    const etype = event.type;
    const data = event.data || {};

    if (etype === "trade") {
        // TradeExecutedEventDto: type_ → wire key "type", value "Market"/"Limit"/"Stop"
        const sideIsBuy = data.side === 0 || data.side === "Buy" || data.side === "buy";
        const otRaw = orderTypeRaw(data.type);
        return {
            symbol: data.symbol,
            side: sideIsBuy ? "BUY" : "SELL",
            orderType: orderTypeLabel(otRaw),
            quantity: Number(data.size || 0),
            price: Number(data.price || 0),
            signalId: data.status === "closed" && data.entryId ? data.entryId : data.tradeId,
            timestamp: Number(data.timestamp || 0),
            eventType: 0,
            orderTypeRaw: otRaw,
            sideRaw: sideIsBuy ? 0 : 1,
        };
    }

    if (etype === "order") {
        // OrderEventDto keys: orderId, orderType, triggerPrice, side, size, timestamp
        const sideIsBuy = data.side === 0 || data.side === "Buy" || data.side === "buy";
        const otRaw = orderTypeRaw(data.orderType ?? data.order_type);
        return {
            symbol: data.symbol,
            side: sideIsBuy ? "BUY" : "SELL",
            orderType: orderTypeLabel(otRaw),
            quantity: Number(data.size || 0),
            price: Number(data.triggerPrice ?? data.trigger_price ?? 0),  // camelCase key
            signalId: data.orderId ?? data.order_id,                      // camelCase key
            timestamp: Number(data.timestamp || 0),
            eventType: 1,
            orderTypeRaw: otRaw,
            sideRaw: sideIsBuy ? 0 : 1,
        };
    }

    if (etype === "alert") {
        // WatchAlert has no side/order_type — defaults to MARKET BUY
        const orderTypeStr = String(data.order_type || data.orderType || "market").toLowerCase();
        const sideStr = String(data.side || "buy").toLowerCase();
        const sideIsBuy = sideStr === "buy" || sideStr === "0";
        const otRaw = orderTypeRaw(orderTypeStr === "limit" ? "Limit" : orderTypeStr === "stop" ? "Stop" : "Market");
        return {
            symbol: data.symbol || "global",
            side: sideIsBuy ? "BUY" : "SELL",
            orderType: orderTypeLabel(otRaw),
            quantity: Number(data.size || 0),
            price: Number(data.price || 0),
            signalId: data.id,
            timestamp: Number(data.timestamp || 0),
            eventType: 2,
            orderTypeRaw: otRaw,
            sideRaw: sideIsBuy ? 0 : 1,
        };
    }

    return null;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
    const configFile = process.env.TICKENGINE_CONFIG_FILE || "config.json";
    console.log(`Loading config from: ${configFile}`);
    const config = JSON.parse(fs.readFileSync(configFile, "utf8"));

    const agents: ExecutionAgent[] = [];

    for (const [agentName, agentCfg] of Object.entries<any>(config.agents || {})) {
        const atype = agentCfg.type;
        const symbolMap = agentCfg.symbol_map || {};

        if (atype === "mt5_ea") {
            const tcpBind = agentCfg.tcp_bind;          // renamed from zmq_bind
            agents.push(new Mt5Agent(agentName, symbolMap, tcpBind));
        } else if (atype === "binance_spot") {
            agents.push(new BinanceAgent(
                agentName,
                symbolMap,
                agentCfg.api_key,
                agentCfg.api_secret,
                agentCfg.api_url || "https://api.binance.com"
            ));
        } else if (atype === "okx_spot") {
            agents.push(new OkxAgent(
                agentName,
                symbolMap,
                agentCfg.api_key,
                agentCfg.api_secret,
                agentCfg.passphrase,
                agentCfg.api_url || "https://www.okx.com"
            ));
        }
    }

    const client = new TickEngineClient(
        config.stream.url,
        config.stream.apiKey || config.stream.api_key,
        config.stream.accountId || config.stream.account_id
    );

    console.log("Connecting to TickEngine stream...");
    client.onEvent(async (event) => {
        const details = extractOrderDetails(event);
        if (!details) return;

        for (const agent of agents) {
            if (agent.canHandle(details.symbol)) {
                try {
                    await agent.execute(event, details);
                } catch (err) {
                    console.error(`Error executing event via Agent '${agent.name}':`, err);
                }
            }
        }
    });
}

main().catch(console.error);
